### ISSUE TITLE
Restore session does not replace current session

### ISSUE THREAD

**Issue by sysread** [ReDEnergy/SessionSync#24]
My expectation as a user is that restoring a session _replaces_ all active tabs, including pinned tabs, with those of the session I wish to load. As an example, I have a saved session that includes some pinned personal email/calendar tabs as well as a number of work-related (unpinned) tabs that I open as a group when I start working in the morning. I often have personal tabs leftover from the previous day and wish to reset to my initial work desktop when I start working.

However, when I load a session, it adds it on top of the current session. There should at least be an option to replace current tabs with the requested session, if not the default behavior.

**Comment by ReDEnergy** [comment]
Hi. For the moment you can use **Restore in new windows** to have a clean restore session. Regarding your preference, I decided not to replace active tabs because closing any tab without user confirmation would be a bad design choice. Of course each of us has different preferences and expectation but the default behavior would be impossible to meet everyone's expectations. As you can say, an option can be added so I will try to add it in the next update.

**Comment by c-e-p-x-u-o** [comment]
It's useless without this feature, as I have a lot of containers.

**Comment by Cerberus-tm** [comment]
Adding it as an option would be nice! For example, this is how Session Manager does it:
![alt text](https://i.imgur.com/fxdy9US.png)

(By "Append to session" it means "create a new window (or several) and restore the session in that window".)

**Comment by dkrm0** [comment]
Same here. To get rid of my current session I need to restore the desired session in a new window and then close the old one right after.

**Comment by cianoz** [comment]
Me too would have this option
Replacing the current window is what I would to do in 99% of cases

**Comment by Cerberus-tm** [comment]
Same here. There could be a confirmation prompt when you replace existing tabs.

**Comment by ReDEnergy** [comment]
Ok.. ok .. Got it... will try to land something ASAP.
The only issue is what would happen if you have multiple windows open ? Should I only consider the one where you are currently accessing the addon or all of them? Because the way I would implement this is to only replace the current window and not close the others as well. Is that ok with you guys or do you really want a clean restore... and all the other windows to be closed automatically ? Though since in the case of this addon 1 saved session is only 1 window, I would say that it might be quite strange to close all the other windows as well. But then restoring from the history might be the other way around... closing just one window, leaving the rest open and opening more then 1 window.
Basically it's not that I don't want to implement this... it's just that I have no idea what exactly meets everyone expectations.

Or should I just open a small confirmation popup with the following options, to be sure everything 
 * close current window and restore the session
 * close all windows and restore the session

Waiting for feedback. Would it be annoying if I put all the options is a confirmation popup? Because doing this leave me the option to provide whatever behaviour you would expect

I will try to land an update version with this by the end of the week!

**Comment by cianoz** [comment]
Thanks for the reply!
IMHO, replacing the current / single window would be a good solution
It would be OK to replace all windows for me, too, although I prefer the 1st option (current window)

**Comment by Cerberus-tm** [comment]
Red Energy, thanks for your very considerate reply! My scenario is as follows:

- I have a few windows open, one with my Gmail, another with a calendar, a third with various other tabs. The windows could be for a work project.
- Then I want to switch to some other activity, like programming. So I need all the old windows to be closed, and I need several new windows opened, like one with my private e-mail, one with various tabs on Javascript documentation, and one with a chat application.

This is fairly common for me, and it is the one situation in which I miss being able to replace one session with another. A situation where I want to keep all windows open, but replace only a single window with another window: this happens very rarely for me. I don't mind doing that manually.

So, for me, the word "session" means all open windows, not just one window. I'd just call that a "window". I admit I don't understand every option in the add-on 100%; I've sometimes found the use of the word "session" confusing, but with time I have come to understand that it sometimes means just one window. I always double check.

Putting all the options in a confirmation pop-up would be fine with me. Alternatively, you could do it the way Session Manager did it: when you click on an old session, you immediately see a couple of options in the existing UI (as in the screenshot above). So you don't actually need to click on an additional pop-up while still being able to choose between options.

At any rate, a pop-up would be fine, a really great improvement over the current situation! Otherwise I love the add-on. I have been using it for about eight months and it has never let me down (the only other little thing that bothers me is the fact that pinned tabs aren't restored as such, but, oh, well).

**Comment by Cerberus-tm** [comment]
P.S. So does that mean something would have to change in the "Sessions" tab of the add-on? Currently, it seems we cannot save a session with more than one window? Or perhaps I misunderstand some of the options.

**Comment by ReDEnergy** [comment]
I will do a popup then.
Regarding the other aspects... Sadly since there is not a really easy way to represent multiple sessions inside the same bookmark folder (actually, I could probably use bookmarks separators for that... it just hit me right now) I always had to remain with the idea that 1 session = 1 window. Of course, that's simply not the case for many users, but it's not 100% say that it's the other way around always. It's just a difference in what we each use windows for. Some people like to have multiple windows that completely separate their work, other just have multiple windows just for convenience or simply because during certain tasks you had to popup a certain tab into a window, have some other documentation elsewhere, but basically everything will remain the same thing. Since I can't do both well I just had to select one option.

Regarding the second message. You can save multiple windows but they will only be saved into separated bookmark folders. But in the case of Auto-saved history, since I can do whatever I want with the structure of memory (not bound by the limitation of bookmarks) it's easier to have a session with multiple windows similar to what Session Manager does (or other managers). It's just that here the bookmarking system offers a lot of good aspects but at the same time is very limited after the WebExtension update. Sadly that's why I was not able to provide pin-tab save information after webextension port simply because Mozilla removed any kind of additional data that can be saved with a bookmark.
But since the core feature of the addon is actually the bookmark saving part, that's for sure won't change ever. Thus, I will try to find various solutions and methods to add missing features in different way, which for sure will always make the feature comparison between this addon and others a little by unfair, but that's all I can do.

With that in mind I think there might be huge issue with the replacing of the session. I can close whatever else is on the screen or the current window... but I can't save them. I have no tracking information about which sessions are opened, new tabs, if there is more than 1 window then there can be multiple "sessions", etc, etc. So saving anything prior to closing everything is really problematic. If I can't do this, will this replace feature still be useful?

**Comment by cianoz** [comment]
In my opinion replacing a single window even without saving anything of it is more than a decent solution. It would cover a very large scenario of use. This is my thought, of course.
And anyway having this feature would be better than having no-feature in this direction.
Luciano Zanin
 

    Il martedì 27 agosto 2019, 08:59:38 CEST, Gabriel Ivanica <notifications@github.com> ha scritto:  
 
 
I will do a popup then.
Regarding the other aspects... Sadly since there is not a really easy way to represent multiple sessions inside the same bookmark folder (actually, I could probably use bookmarks separators for that... it just hit me right now) I always had to remain with the idea that 1 session = 1 window. Of course, that's simply not the case for many users, but it's not 100% say that it's the other way around always. It's just a difference in what we each use windows for. Some people like to have multiple windows that completely separate their work, other just have multiple windows just for convenience or simply because during certain tasks you had to popup a certain tab into a window, have some other documentation elsewhere, but basically everything will remain the same thing. Since I can't do both well I just had to select one option.

Regarding the second message. You can save multiple windows but they will only be saved into separated bookmark folders. But in the case of Auto-saved history, since I can do whatever I want with the structure of memory (not bound by the limitation of bookmarks) it's easier to have a session with multiple windows similar to what Session Manager does (or other managers). It's just that here the bookmarking system offers a lot of good aspects but at the same time is very limited after the WebExtension update. Sadly that's why I was not able to provide pin-tab save information after webextension port simply because Mozilla removed any kind of additional data that can be saved with a bookmark.
But since the core feature of the addon is actually the bookmark saving part, that's for sure won't change ever. Thus, I will try to find various solutions and methods to add missing features in different way, which for sure will always make the feature comparison between this addon and others a little by unfair, but that's all I can do.

With that in mind I think there might be huge issue with the replacing of the session. I can close whatever else is on the screen or the current window... but I can't save them. I have no tracking information about which sessions are opened, new tabs, if there is more than 1 window then there can be multiple "sessions", etc, etc. So saving anything prior to closing everything is really problematic. If I can't do this, will this replace feature still be useful?

**Comment by ReDEnergy** [comment]
Fair point. Then, a first version will be implemented by the end of the week as I promised.

**Comment by ReDEnergy** [comment]
Sorry about the delay. Did not forget about this, but I was not able to work on it to much till today. I got a version that offers 2 options: **restore (close active)** and **restore (close all)**.
If anyone can please install the beta version and send me some feedback that would be great

[Download Beta](https://www.dropbox.com/s/3gijyuwo0mth46w/session_sync-3.1.13b-fx.xpi?dl=0)

The beta is just the release with the added options so it's 100% safe to use. When I will release the final version it will be automatically updated because I always make sure I increase the version number so anyone that wants to start using it right now can do that.

Regarding icons/namings/tooltips for the new options. Yes I'm totally aware that they may not be ok, but I simply have no idea what to chose for the moment. I will think more about them for the actual release. Probably I will add some option to hide certain toolbar button because there are to many and I bet very few users actually use all of them.

I'm 100% open to suggestion but please have in mind that somehow I have to make sure things feel and look ok for everyone not just for one person so don't mind if I decide not to follow certain personal preferences.

**Comment by dkrm0** [comment]
I have just installed it. It works great :)
